<!DOCTYPE HTML>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
    <title></title>
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="description" content="">
    <link rel="stylesheet" href="gitbook/style.css">
    <link rel="stylesheet" href="gitbook/website.css">
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="favicon.ico" type="image/x-icon">
  </head>
  <body>
    <div class="book">
      <div class="book-summary">
        <div id="book-search-input" role="search">
          <input type="text" placeholder="Type to search" />
        </div>
        
        <nav class="summary" id="TOC" role="doc-toc">
          <h1 id="summary">Summary</h1>
          <ul>
          <li><a href="README.html">Introduction</a></li>
          <li><a href="pigx-rna-seq.html">PiGx RNA-seq</a>
          <ul>
          <li><a href="pigx-rna-seq.html#introduction">Introduction</a></li>
          <li><a href="pigx-rna-seq.html#installation">Installation</a></li>
          <li><a href="pigx-rna-seq.html#quick-start">Quick start</a></li>
          <li><a href="pigx-rna-seq.html#preparing-the-input">Preparing the input</a></li>
          <li><a href="pigx-rna-seq.html#running-the-pipeline">Running the pipeline</a></li>
          <li><a href="pigx-rna-seq.html#output-description">Output description</a></li>
          <li><a href="pigx-rna-seq.html#troubleshooting">Troubleshooting</a></li>
          <li><a href="pigx-rna-seq.html#questions">Questions</a></li>
          </ul></li>
          <li><a href="pigx-chip-seq.html#pigx-chip-seq">PiGx ChIP-seq</a>
          <ul>
          <li><a href="pigx-chip-seq.html#introduction">Introduction</a></li>
          <li><a href="pigx-chip-seq.html#installation">Installation</a></li>
          <li><a href="pigx-chip-seq.html#quick-start">Quick Start</a></li>
          <li><a href="pigx-chip-seq.html#preparing-input">Preparing Input</a></li>
          <li><a href="pigx-chip-seq.html#running-the-pipeline">Running the Pipeline</a></li>
          <li><a href="pigx-chip-seq.html#output-description">Output Description</a></li>
          <li><a href="pigx-chip-seq.html#troubleshooting">Troubleshooting</a></li>
          <li><a href="pigx-chip-seq.html#questions">Questions</a></li>
          </ul></li>
          <li><a href="pigx-bs-seq.html">PiGx BS-seq</a>
          <ul>
          <li><a href="pigx-bs-seq.html#introduction">Introduction</a></li>
          <li><a href="pigx-bs-seq.html#installation">Installation</a></li>
          <li><a href="pigx-bs-seq.html#quick-start">Quick start</a></li>
          <li><a href="pigx-bs-seq.html#preparing-the-input">Preparing the input</a></li>
          <li><a href="pigx-bs-seq.html#running-the-pipeline">Running the pipeline</a></li>
          <li><a href="pigx-bs-seq.html#output-description">Output description</a></li>
          <li><a href="pigx-bs-seq.html#troubleshooting">Troubleshooting</a></li>
          <li><a href="pigx-bs-seq.html#questions">Questions</a></li>
          </ul></li>
          <li><a href="pigx-scrna-seq.html">PiGx scRNA-seq</a>
          <ul>
          <li><a href="pigx-scrna-seq.html#introduction">Introduction</a></li>
          <li><a href="pigx-scrna-seq.html#installation">Installation</a></li>
          <li><a href="pigx-scrna-seq.html#quick-start">Quick start</a></li>
          <li><a href="pigx-scrna-seq.html#preparing-the-input">Preparing the input</a></li>
          <li><a href="pigx-scrna-seq.html#running-the-pipeline">Running the pipeline</a></li>
          <li><a href="pigx-scrna-seq.html#output-description">Output description</a></li>
          <li><a href="pigx-scrna-seq.html#troubleshooting">Troubleshooting</a></li>
          <li><a href="pigx-scrna-seq.html#questions">Questions</a></li>
          </ul></li>
          <li><a href="pigx-sars-cov2-ww.html">PiGx SARS-CoV-2 wastewater</a>
          <ul>
          <li><a href="pigx-sars-cov2-ww.html#introduction">Introduction</a></li>
          <li><a href="pigx-sars-cov2-ww.html#installation">Installation</a></li>
          <li><a href="pigx-sars-cov2-ww.html#quick-start">Quick start</a></li>
          <li><a href="pigx-sars-cov2-ww.html#preparing-the-input">Preparing the input</a></li>
          <li><a href="pigx-sars-cov2-ww.html#troubleshooting">Troubleshooting</a></li>
          </ul></li>
          </ul>
        </nav>
      </div>

      <div class="book-body">
        <div class="body-inner">
          <div class="book-header" role="navigation">
            <!-- Title -->
            <h1>
              <i class="fa fa-circle-o-notch fa-spin"></i>
              <a href="." ></a>
            </h1>
          </div>
          <div class="page-wrapper" tabindex="-1" role="main">
            <div class="page-inner">
              <div id="book-search-results">
                <div class="search-noresults">
                  <section class="normal markdown-section">
                    <h1 id="pigx-sars-cov-2-wastewater">PiGx SARS-CoV-2 wastewater</h1>
                    <h1 id="introduction">Introduction</h1>
                    <p>PiGx SARS-CoV-2 is a pipeline for analysing data from sequenced wastewater samples and identifying given variants-of-concern of SARS-CoV-2. Currently wastewater samples are used, which are enriched for SARS-CoV-2. The pipeline can be used for continuous sampling. The output of the PiGx SARS-CoV-2 pipeline is summarized in a report which provides an intuitive visual overview about the development of variant abundance over time and location. Additionally there will be more detailed reports per sample, which cover the quality control of the samples, the detected variants and a taxonomic classification of all unaligned reads. This version of the pipeline was designed to work with target sequencing (PCR-based) and tested with data generated using the <a href="https://github.com/artic-network/artic-ncov2019/tree/master/primer_schemes/nCoV-2019/V30">ARTIC nCoV-2019 primers</a>.</p>
                    <h2 id="workflow">Workflow</h2>
                    <p>First the raw reads are trimmed by using <a href="http://prinseq.sourceforge.net/">Prinseq</a> to improve alignment rates and mutation calling. Next, the trimmed reads are aligned to the reference genome of SARS-CoV-2 using <a href="https://github.com/lh3/bwa">BWA</a>, and the results are <em>SAM</em>/<em>BAM</em> files of <strong>aligned</strong> and <strong>unaligned reads</strong>. Following the alignment a quality check on raw and processed reads is performed by using <a href="https://multiqc.info/">MultiQC</a>. Calling the variants and inferring SNVs (single nucleotide polymorphisms) on the <strong>aligned reads</strong> is done with <a href="https://csb5.github.io/lofreq/">LoFreq</a>. Mutations are annotated with <a href="https://covid-19.ensembl.org/index.html">VEP</a>. Estimation of Variants of Concern (VOC) frequencies are done by deconvolution. To investigate the abundance of other existing species in the wastewater samples the <strong>unaligned reads</strong> will be taxonomicly classified with <a href="https://github.com/DerrickWood/kraken2">Kraken2</a>. The Kraken2 requires a database of the genomes against the reads are getting aligned, therefore keep in mind that you can only find those species which are included in the chosen database. For documentation how to set this up, see: <a href="#prepare-databases">Prepare databases</a>. For a better and interactive visualization of all species present in the wastewater <a href="https://github.com/marbl/Krona/wiki">Krona</a> is used. Also here a small step of setting up a database is needed before running the pipeline, see: <a href="#prepare-databases">Prepare databases</a>.</p>
                    <h2 id="output">Output</h2>
                    <ul>
                    <li>overview report including:
                    <ul>
                    <li>Visualization of the development of SARS-CoV-2 variants and mutations over time and locations from all samples provided.</li>
                    <li>Quality Control report per sample: number of covered amplicons, read coverage and MultiQC report for raw and trimmed reads</li>
                    <li>Variants report per sample: variant analysis of SARS-CoV-2 from each wastewater sample and identification of variants of concern</li>
                    <li>Taxonomic classification: a table and pie chart of the species found in the unaligned reads</li>
                    </ul></li>
                    <li><em>SAM</em> / <em>BAM</em> files per sample: aligned and unaligned reads against SARS-CoV-2</li>
                    <li><em>VCF</em> / <em>CSV</em> files per sample: listing all detected single nucleotide variants (SNVs) from the aligned reads</li>
                    <li>VEP reports per sample as <em>TXT</em> / <em>HTML</em>: report files show the <a href="https://www.ensembl.org/info/docs/tools/vep/vep_formats.html#output">VEP output</a> including the uploading variants, consequenting amino acid changes and consequences for the corresponding protein</li>
                    <li>Kraken2 files per sample <em>(txt)</em>: provides an overview of all found species in the unaligned reads together with the NCBI taxonomy ID</li>
                    <li>log files for all major analysis steps performed by the pipeline</li>
                    </ul>
                    <h1 id="installation">Installation</h1>
                    <p>Pre-built binaries for PiGx are available through <a href="https://gnu.org/s/guix">GNU Guix</a>, the functional package manager for reproducible, user-controlled software management. You can install the PiGx SARS-CoV-2 pipeline with</p>
                    <div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a><span class="ex">guix</span> install pigx-sars-cov2-ww</span></code></pre></div>
                    <p>If you want to install PiGx SARS-CoV-2 from source, please clone this repository and change directory accordingly:</p>
                    <div class="sourceCode" id="cb2"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a><span class="fu">git</span> clone https://github.com/BIMSBbioinfo/pigx_sarscov2_ww.git</span>
                    <span id="cb2-2"><a href="#cb2-2" aria-hidden="true"></a><span class="bu">cd</span> pigx_sarscov2_ww</span></code></pre></div>
                    <p>To fetch code that is common to all PiGx pipelines run this:</p>
                    <div class="sourceCode" id="cb3"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a><span class="fu">git</span> submodule update --init</span></code></pre></div>
                    <p>Before setting everything up, though, make sure <a href="https://github.com/BIMSBbioinfo/pigx_sarscov2_ww/blob/main/manifest.scm">all dependencies</a> are met by either installing them manually, or by entering the provided reproducible Guix environment. If you are using Guix we definitely recommend the latter. This command spawns a sub-shell in which all dependencies are available at exactly the same versions that we used to develop the pipeline:</p>
                    <div class="sourceCode" id="cb4"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true"></a><span class="va">USE_GUIX_INFERIOR=</span>t <span class="ex">guix</span> environment --pure -m manifest.scm --preserve=GUIX_LOCPATH</span></code></pre></div>
                    <p>To use your current Guix channels instead of the fixed set of channels, just omit the <code>USE_GUIX_INFERIOR</code> shell variable:</p>
                    <div class="sourceCode" id="cb5"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true"></a><span class="ex">guix</span> environment --pure -m manifest.scm --preserve=GUIX_LOCPATH</span></code></pre></div>
                    <p>Note that <code>--pure</code> unsets all environment variables that are not explicitly preserved. To access other executables that are not part of the environment please address them by their absolute file name.</p>
                    <p>Inside the environment you can then perform the usual build steps:</p>
                    <div class="sourceCode" id="cb6"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true"></a><span class="ex">./bootstrap.sh</span></span>
                    <span id="cb6-2"><a href="#cb6-2" aria-hidden="true"></a><span class="ex">./configure</span> --prefix=/some/where</span>
                    <span id="cb6-3"><a href="#cb6-3" aria-hidden="true"></a><span class="fu">make</span> install</span></code></pre></div>
                    <h1 id="getting-started">Getting started</h1>
                    <p>At this point you are able to run PiGx SARS-CoV-2. To see all available options type <code>--help</code>.</p>
                    <div class="sourceCode" id="cb7"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true"></a><span class="ex">pigx-sars-cov2-ww</span> --help</span></code></pre></div>
                    <details>
                    <summary>toggle output</summary>

                    <pre>
                    <code>
                    usage: pigx-sars-cov2-ww [-h] [-v] (--init [{settings,sample-sheet,both}] | -s SETTINGS) [-c CONFIGFILE]
                                             [--target TARGET] [-n] [--graph GRAPH] [--force] [--reason] [--unlock] [--verbose]
                                             [--printshellcmds]
                                             [sample_sheet]

                    PiGx SARS-CoV-2 wastewater sequencing Pipeline.

                    This is a pipeline for analyzing data from sequenced wastewater
                    samples and identifying given variants-of-concern of SARS-CoV-2.  The
                    pipeline can be used for continuous sampling.  The output report
                    provides an intuitive visual overview about the development of variant
                    abundance over time and location.

                    positional arguments:
                      sample_sheet                            The sample sheet containing sample data in CSV format.

                    optional arguments:
                      -h, --help                              show this help message and exit
                      -v, --version                           show program's version number and exit
                      --init [{settings,sample-sheet,both}]   Generate a template SETTINGS file, a SAMPLE-SHEET.  Leave empty for both.
                      -s SETTINGS, --settings SETTINGS        A YAML file for settings that deviate from the defaults.
                      -c CONFIGFILE, --configfile CONFIGFILE  The config file used for calling the underlying snakemake process.  By
                                                              default the file 'config.json' is dynamically created from the sample
                                                              sheet and the settings file.
                      --target TARGET                         Stop when the named target is completed instead of running the whole
                                                              pipeline.  The default target is "final-report".  Pass "--target=help"
                                                              to describe all available targets.
                      -n, --dry-run                           Only show what work would be performed.  Do not actually run the
                                                              pipeline.
                      --graph GRAPH                           Output a graph in PDF format showing the relations between rules of
                                                              this pipeline.  You must specify a graph file name such as
                                                              "graph.pdf".
                      --force                                 Force the execution of rules, even though the outputs are considered
                                                              fresh.
                      --reason                                Print the reason why a rule is executed.
                      --unlock                                Recover after a snakemake crash.
                      --verbose                               Print supplementary info on job execution.
                      --printshellcmds                        Print commands being executed by snakemake.

                    This pipeline was developed by the Akalin group at MDC in Berlin in 2017-2021.
                    </code>
                    </pre>
                    </details>
                    <p></br></p>
                    <p>Though, to actually use it on your experimental data still more setup is required. Please follow <a href="#prepare-databases">the steps to prepare the required databases</a> first and <a href="#preparing-the-input">prepare the input</a>. Then afterwards, you can run the pipeline:</p>
                    <div class="sourceCode" id="cb8"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true"></a><span class="ex">pigx-sars-cov2-ww</span> [options] sample_sheet.csv</span></code></pre></div>
                    <p>For example, with this command the pipeline is used for the included test data:</p>
                    <div class="sourceCode" id="cb9"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true"></a><span class="ex">pigx-sars-cov2-ww</span> -s tests/settings.yaml tests/sample_sheet.csv</span></code></pre></div>
                    <h1 id="preparing-the-input">Preparing the input</h1>
                    <p>In order to run the pipeline, you need to supply</p>
                    <ul>
                    <li>a sample sheet</li>
                    <li>a settings file.</li>
                    </ul>
                    <p>Both files are described below.</p>
                    <p>In order to generate template settings and sample sheet files, type</p>
                    <div class="sourceCode" id="cb10"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true"></a><span class="ex">pigx-sars-cov2-ww</span> --init</span></code></pre></div>
                    <p>in the shell, and a boilerplate <code>sample_sheet.csv</code> and <code>settings.yaml</code> will be written to your current directory. An example for both files is provided in the <code>tests/</code> directory.</p>
                    <h2 id="sample-sheet">Sample sheet</h2>
                    <p>The sample sheet is a tabular file (<code>csv</code> format) describing the experiment. The table has the following columns:</p>
                    <table style="width:100%;">
                    <colgroup>
                    <col style="width: 11%" />
                    <col style="width: 13%" />
                    <col style="width: 15%" />
                    <col style="width: 15%" />
                    <col style="width: 15%" />
                    <col style="width: 15%" />
                    <col style="width: 15%" />
                    </colgroup>
                    <thead>
                    <tr class="header">
                    <th>SampleName</th>
                    <th>Read</th>
                    <th>Read2</th>
                    <th>date</th>
                    <th>location_name</th>
                    <th>coordinates_lat</th>
                    <th>coordinates_long</th>
                    </tr>
                    </thead>
                    <tbody>
                    <tr class="odd">
                    <td>Test0</td>
                    <td>Test0_R1.fastq</td>
                    <td>Test0_R2.fastq</td>
                    <td>2021-01-01T08:00:00</td>
                    <td>Berlin</td>
                    <td>52.3646054650197</td>
                    <td>13.5098643274129</td>
                    </tr>
                    <tr class="even">
                    <td>Test2</td>
                    <td>Test2_R1.fastq</td>
                    <td>Test2_R2.fastq</td>
                    <td>2021-01-03T08:00:00</td>
                    <td>Munich</td>
                    <td>48.2084486780314</td>
                    <td>11.6282300407931</td>
                    </tr>
                    </tbody>
                    </table>
                    <ul>
                    <li><em>SampleName</em> is the name for the sample</li>
                    <li><em>Read</em> &amp; <em>Read2</em> are the fastq file names of paired end reads
                    <ul>
                    <li>the location of these files is specified in the settings file</li>
                    <li>single-end data is not yet supported</li>
                    <li>compressed (<code>.gz</code>) reads are not yet supported</li>
                    </ul></li>
                    <li><em>date</em> is a date/time in ISO format (<code>yyyy-mm-ddThh:mm:ss</code>)</li>
                    <li><em>location_name</em> is the name of the location and should be unique per coordinates</li>
                    <li><em>coordinates_lat</em> &amp; <em>coordinates_long</em> correspond the latitude and longitude of the location name</li>
                    </ul>
                    <h2 id="settings-file">Settings file</h2>
                    <p>The settings file contains parameters (in YAML format) to configure the execution of the PiGx SARS-CoV-2 pipeline. It specifies:</p>
                    <p><strong>Locations</strong>:</p>
                    <ul>
                    <li><em>output-dir</em>, the location of the outputs for the pipeline</li>
                    <li><em>reads-dir</em>, the location of the reads (directory where <code>fastq</code> files are)</li>
                    <li><em>reference-fasta</em>, the <code>fasta</code> file with the reference genome (must be prepared by the user)</li>
                    <li><em>amplicons-bed</em>, the amplicons <code>bed</code> file for coronavirus (must be prepared by the user)</li>
                    <li><em>kraken-db-dir</em>, the location of the kraken database (must be prepared by the user)</li>
                    <li><em>krona-db-dir</em>, the location of the krona database (must be prepared by the user)</li>
                    <li><em>sigmut-db-dir</em>, the location of the signature mutations database (provided at databases/sigmut_db/)</li>
                    <li><em>vep-db-dir</em>, the location of <code>sars_cov_2</code> database for VEP (must be prepared by the user)</li>
                    </ul>
                    <p><strong>Trimming</strong>:</p>
                    <p>These settings are used to filter raw reads when trimming. Reads that are shorter than the product of read-length and the cut-off factor are removed.</p>
                    <ul>
                    <li><em>read-length</em> specifies the length of the basepairs</li>
                    <li><em>cut-off</em> specifies the cut-off factor</li>
                    </ul>
                    <h2 id="prepare-databases">Prepare databases</h2>
                    <p>Before the pipeline can work, three databases must be downloaded and their location will need to be provided in the settings file. Depending on the size of the databases this can take some time. Be sure that the pigx-sarscov2-ww pipeline is downloaded and the tools are installed or used via the provided and suggested guix environment. One database (signature mutations, <code>sigmut_db</code>) is already provided via the repository. The directory structure is suggested like <a href="#structure-overview">this</a> and pre-filled accordingly in the settings file.</p>
                    <h3 id="kraken2-database">Kraken2 database</h3>
                    <p>There are several libraries of genomes that can be used to classify the (unaligned) reads. It is up to you which one to use, but be sure that they fulfill the necessities stated by Kraken2 <a href="https://github.com/DerrickWood/kraken2/wiki/Manual#kraken-2-databases">Kraken2 manual</a>. For an overall overview we recommend to use the Plus-PFP library provided <a href="https://benlangmead.github.io/aws-indexes/k2">here</a>. If the classification is not of concern or only the viruses are of interest, we recommend to use a smaller one. This will accelerate the speed. It is also possible to have multiple Kraken2 databases installed, just be sure to provide the correct location to the settings file.</p>
                    <p>First download and unpack the database in the <code>databases/kraken_db/</code>:</p>
                    <pre><code>DIR=databases/kraken_db/
                    mkdir -p $DIR

                    # NOTE: This command will download a very large file, after unpacking this will
                    # require about 100GB of disc space. If this is not feasible use another
                    # database instead. For this please see link above or commented lines below.
                    wget -qO- https://genome-idx.s3.amazonaws.com/kraken/k2_pluspfp_20210127.tar.gz | tar -C $DIR -xzv

                    # Use the following two lines to use smaller 8GB version instead.
                    # wget -qO- https://genome-idx.s3.amazonaws.com/kraken/k2_pluspfp_8gb_20210127.tar.gz | tar -C $DIR -xzv</code></pre>
                    <p>Next go to the database/ directory and build the Kraken database. This might take a while (depends on the size of the downloaded database):</p>
                    <pre><code>#current location: pigx_sarscov2_ww/
                    cd database/
                    DBNAME=kraken_db
                    kraken2-build --use-ftp --download-taxonomy --db $DBNAME # if this fails, you might want to try it without the --use-ftp flag
                    kraken2-build --build --db $DBNAME</code></pre>
                    <p>Kraken might tell you that it can’t find a library subdirectory. If that’s the case it should be fine though.<br />
                    This folder should now contain at least these files: hash.k2d, opts.k2d and taxo.k2d.</p>
                    <h3 id="krona-database">Krona database</h3>
                    <p>Krona Tools needs two files, which have to be installed in the <code>databases/krona_db/</code>. Also this might take a while:</p>
                    <pre><code>#current location: pigx_sarscov2_ww/databases/
                    DBNAME=krona_db/
                    mkdir -p $DBNAME
                    KRONA=$(dirname $(which ktImportTaxonomy))/../share/krona-tools/ # this is just a workaround until tool paths are declared
                    $KRONA/updateTaxonomy.sh $DBNAME # the scripts are stored a priori in that folder
                    $KRONA/updateAccessions.sh $DBNAME</code></pre>
                    <h3 id="vep-database">VEP database</h3>
                    <p>Just download the <code>SARS_CoV_2</code> database for VEP (variant effect predictor) and unpack it in the <code>databases/vep_db/</code> directory.</p>
                    <pre><code>#current location: pigx_sarscov2_ww/databases/
                    DBNAME=vep_db/
                    mkdir -p $DBNAME
                    wget -qO- ftp://ftp.ensemblgenomes.org/pub/viruses/variation/indexed_vep_cache/sars_cov_2_vep_101_ASM985889v3.tar.gz | tar -C $DBNAME -xzv</code></pre>
                    <h3 id="sigmut-database">Sigmut database</h3>
                    <p>Necessary files are provided in <code>databases/sigmut_db/</code> for the current main Variants of Concern. Users can add files with new variants if/when necessary.</p>
                    <h1 id="quick-start">Quick Start</h1>
                    <p>To check wether the pipeline together with the databases was properly installed, run PiGx SARS-CoV-2 Wastewater Sequencing Pipeline on a minimal test dataset. For this there are samples provided in <code>tests/sample_data/</code>. The directory structure should be provided like this, assuming all databases are set up like described <a href="#prepare-databases">here</a>:</p>
                    <pre><code>pigx_sarscov2_ww
                    │
                    ├── databases
                    │   ├── kraken_db
                    │   │   └── ...
                    │   ├── krona_db
                    │   │   └── ...
                    │   ├── sigmut_db
                    │   │   └── ...
                    │   └── vep_db
                    │       └── ...
                    ├── tests
                    │   ├── databases
                    │   │   └── vep_db
                    │   │       └── ...
                    │   ├── sample_data
                    │   │   ├── reads
                    │   │   │   ├── ...
                    │   │   ├── signature_mutations
                    │   │   │   └── ...
                    │   │   ├── NC_045512.2.fasta
                    │   │   ├── nCoV-2019_NCref.bed
                    │   │   └── ...
                    │   ├── sample_sheet.csv
                    │   ├── settings.yaml
                    │   └── ...
                    └── ...</code></pre>
                    <p>Now the the test set can be run with the command:</p>
                    <pre><code>#current location: pigx_sarscov2_ww/
                    PIGX_UNINSTALLED=t ./pigx-sars-cov2-ww -s tests/settings.yaml tests/sample_sheet.csv</code></pre>
                    <p>Inside <code>tests/</code> a new directory <code>output</code> is created, which includes specific directories containing output data for the respective step of the pipeline. The <code>tests/output/reports/index.html</code> gives the overview over all merged reports for the test data.</p>
                    <h1 id="troubleshooting">Troubleshooting</h1>
                    <p>If you have any questions please e-mail: <a href="mailto:pigx@googlegroups.com" class="email">pigx@googlegroups.com</a> or use the web form to ask questions <a href="https://groups.google.com/forum/#!forum/pigx/" class="uri">https://groups.google.com/forum/#!forum/pigx/</a>.</p>
                    <p>If you run into any bugs, please open an issue here: <a href="https://github.com/BIMSBbioinfo/pigx_rnaseq/issues" class="uri">https://github.com/BIMSBbioinfo/pigx_rnaseq/issues</a>.</p>
                  </section>
                </div>
                <div class="search-results">
                  <div class="has-results">
                    <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
                    <ul class="search-results-list"></ul>
                  </div>
                  <div class="no-results">
                    <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <script>
        var gitbook = gitbook || [];
        gitbook.push(function() {
          gitbook.page.hasChanged(
            {
              "page":
              {
                "title":"",
                "level":"1.1",
                "depth":1,
                "next":{},
                "dir":"ltr"
              },
              "config":
              {
                "gitbook":"*",
                "theme":"default",
                "variables":{},
                "plugins":[],
                "pluginsConfig":
                {
                  "highlight":{},
                  "search":{},
                  "lunr": {
                    "maxIndexSize":1000000,
                    "ignoreSpecialCharacters":false
                  },
                  "sharing": {
                    "facebook":true,
                    "twitter":true,
                    "google":false,
                    "weibo":false,
                    "instapaper":false,
                    "vk":false,
                    "all":[
                      "facebook",
                      "google",
                      "twitter",
                      "weibo",
                      "instapaper"
                    ]},
                  "fontsettings": {
                    "theme":"white",
                    "family":"sans",
                    "size":2
                  },
                  "theme-default": {
                    "styles": {
                      "website":"styles/website.css",
                      "pdf":"styles/pdf.css",
                      "epub":"styles/epub.css",
                      "mobi":"styles/mobi.css",
                      "ebook":"styles/ebook.css",
                      "print":"styles/print.css"
                    },
                    "showLevel":false
                  }
                },
                "structure": {
                  "langs":"LANGS.md",
                  "readme":"README.md",
                  "glossary":"GLOSSARY.md",
                  "summary":"SUMMARY.md"
                },
                "pdf": {
                  "pageNumbers":true,
                  "fontSize":12,
                  "fontFamily":"Arial",
                  "paperSize":"a4",
                  "chapterMark":"pagebreak",
                  "pageBreaksBefore":"/",
                  "margin": {
                    "right":62,"left":62,"top":56,"bottom":56
                  }
                },
                "styles": {
                  "website":"styles/website.css",
                  "pdf":"styles/pdf.css",
                  "epub":"styles/epub.css",
                  "mobi":"styles/mobi.css",
                  "ebook":"styles/ebook.css",
                  "print":"styles/print.css"
                }
              },
              "file": {
                "path":"",
                "mtime":"",
                "type":"markdown"
              },
              "gitbook": {
                "version":"3.2.3",
                "time":"2019-07-08T15:24:30.549Z"
              },
              "basePath":".",
              "book": {"language":""}});
        });
    </script>
    </div>
    <script src="https://code.jquery.com/jquery-3.5.1.min.js"></script>
    <script src="gitbook/gitbook.js"></script>
    <script src="gitbook/theme.js"></script>
    <script src="gitbook/gitbook-plugin-search/search-engine.js"></script>
    <script src="gitbook/gitbook-plugin-search/search.js"></script>
    <script src="gitbook/gitbook-plugin-lunr/lunr.js"></script>
    <script src="gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
    <script src="gitbook/gitbook-plugin-sharing/buttons.js"></script>
    <script src="gitbook/gitbook-plugin-fontsettings/fontsettings.js"></script>
  </body>
</html>
